source/bdDetect.py

```diff
@@ -0,0 +1,370 @@
+#bdDetect.py
+#A part of NonVisual Desktop Access (NVDA)
+#This file is covered by the GNU General Public License.
+#See the file COPYING for more details.
+#Copyright (C) 2013-2017 NV Access Limited
+
+"""Support for braille display detection.
+This allows devices to be automatically detected and used when they become available,
+as well as providing utilities to query for possible devices for a particular driver.
+To support detection for a driver, devices need to be associated
+using the C{add*} functions.
+Drivers distributed with NVDA do this at the bottom of this module.
+For drivers in add-ons, this must be done in a global plugin.
+"""
+
+import itertools
+from collections import namedtuple, defaultdict
+import threading
+import wx
+import hwPortUtils
+import braille
+import winKernel
+import core
+import ctypes
+
+#: How often (in ms) to poll for Bluetooth devices.
+POLL_INTERVAL = 5000
+
+_driverDevices = {}
+
+class DeviceMatch(
+	namedtuple("DeviceMatch", ("type","id", "port", "deviceInfo"))
+):
+	"""Represents a detected device.
+	@ivar id: The identifier of the device.
+	@type id: unicode
+	@ivar port: The port that can be used by a driver to communicate with a device.
+	@type port: unicode
+	@ivar deviceInfo: all known information about a device.
+	@type deviceInfo: dict
+	"""
+	__slots__ = ()
+
+# Device type constants
+#: Key constant for HID devices
+KEY_HID = "hid"
+#: Key for serial devices (com ports)
+KEY_SERIAL = "serial"
+#: Key for devices with a manufacturer specific driver
+KEY_CUSTOM = "custom"
+#: Key foor bluetooth devices
+KEY_BLUETOOTH = "bluetooth"
+
+def _getDriver(driver):
+	try:
+		return _driverDevices[driver]
+	except KeyError:
+		ret = _driverDevices[driver] = defaultdict(set)

```

Comment:
This is @Jcsteh's code, but I kept it this way because it is not only an intermediate variable, but it also adds the driver to the drivers dictionary and sets the value as a DefaultDict(set). The latter is something I changed, as it was a regular dict before.

Note that _getDriver is only used while adding USB and Bluetooth info, which requires a driver specific dictionary to exist.